### PR TITLE
gui: add aggregate net descriptor

### PR DESCRIPTION
**Pull request by gadfort**
https://github.com/The-OpenROAD-Project/OpenROAD/issues/2781

### CODE REVIEW

**Comment by maliberty** [comment]
This looks like a bug as the aggregate is VDD:
image

File: src/gui/src/aggregateNetDescriptor.cpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+  for (auto* iterm : net->getITerms()) {
+    auto* inst = iterm->getInst();
+    if (!AggregateNet::isAggregate(inst)) {
+      continue;
+    }
+
+    for (auto* next_iterm : inst->getITerms()) {
+      populate(next_iterm->getNet());
+    }
```
Comment (maliberty) [inline src/gui/src/aggregateNetDescriptor.cpp:81]: I expect the bug is here in not checking whether the iterm is power/gnd

Comment (gadfort) [reply src/gui/src/aggregateNetDescriptor.cpp:81]: Yeah, you are correct, supplies shouldn't be in the list.

Comment (maliberty) [reply src/gui/src/aggregateNetDescriptor.cpp:81]: if (!next_iterm->getSigType().isSupply()) {
        populate(next_iterm->getNet());
      }

seems to work

**Comment by maliberty** [comment]
After the discussion in the issue I did come to like "Buffer tree" better than "Aggregate Net".   Perhaps it could include the buffer instances as well as the nets.

**Comment by maliberty** [comment]
Also the non-buffer iterm/bterms would be nice (like a regular net) to see where the tree terminates.

**Comment by gadfort** [comment]
@maliberty I renamed it to Buffer Tree, added iterms/bterms and instances to the descriptor (they are not part of the highlight though). Route guides and focus nets is also there.

File: src/gui/src/dbDescriptors.cpp [hunk 2]
```diff
@@ -1332,6 +1333,10 @@ Descriptor::Properties DbNetDescriptor::getProperties(std::any object) const
     props.push_back({"Non-default rule", gui->makeSelected(ndr)});
   }
 
+  if (BufferTree::isAggregate(net)) {
+    props.push_back({"Aggregate net", gui->makeSelected(BufferTree(net))});
```
Comment (maliberty) [inline src/gui/src/dbDescriptors.cpp:1337]: There is still some use of Aggregate in the code and visible in the GUI

File: src/gui/src/bufferTreeDescriptor.cpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+BufferTree::BufferTree(odb::dbNet* net)
+{
+  populate(net);
+
+  std::sort(
+      nets_.begin(), nets_.end(), [](odb::dbNet* r, odb::dbNet* l) -> bool {
+        return r->getName() < l->getName();
+      });
+
+  name_ = nets_[0]->getName();
```
Comment (maliberty) [inline src/gui/src/bufferTreeDescriptor.cpp:58]: std::min_element is more efficient than sort

File: src/gui/src/bufferTreeDescriptor.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+  for (auto* iterm : net->getITerms()) {
+    auto* inst = iterm->getInst();
+    if (!BufferTree::isAggregate(inst)) {
+      if (std::find(iterm_terms_.begin(), iterm_terms_.end(), iterm)
+          == iterm_terms_.end()) {
+        iterm_terms_.push_back(iterm);
+      }
+      continue;
+    }
```
Comment (maliberty) [inline src/gui/src/bufferTreeDescriptor.cpp:88]: A std::set would be more efficient than linear search of a vector.

File: src/gui/src/bufferTreeDescriptor.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    return Selected(*anet, this);
+  }
+  return Selected();
+}
+
+bool BufferTreeDescriptor::lessThan(std::any l, std::any r) const
+{
+  BufferTree* l_anet = std::any_cast<BufferTree>(&l);
+  BufferTree* r_anet = std::any_cast<BufferTree>(&r);
+  return l_anet->getName() < r_anet->getName();
```
Comment (maliberty) [inline src/gui/src/bufferTreeDescriptor.cpp:238]: anet needs updating

**Review by maliberty — CHANGES_REQUESTED** [review]
I am really finding this helpful.  Thanks!

**Comment by maliberty** [comment]
Free bonus, If you click a clock net you get the clock tree:
image

File: src/gui/src/bufferTreeDescriptor.cpp (hunk 3 above)
Comment (gadfort) [reply src/gui/src/bufferTreeDescriptor.cpp:58]: The sort is used to ensure a stable ordering of the nets.

**Comment by gadfort** [comment]
Feedback should be addressed

**Review by maliberty — APPROVED** [review]

**Pull request merged**